deploy/docker-compose/README.md

```diff
@@ -0,0 +1,177 @@
+// failing tasks negates iterator values, shuffle data of failing task must not leak into next stage
+case class FaultyIterator(it: Iterator[java.lang.Long], fail: Boolean, sleep: Option[Int]) extends Iterator[java.lang.Long] {
+  override def hasNext: Boolean = it.hasNext || fail
+  override def next(): java.lang.Long = {
+    // delay iteration if requested
+    if (sleep.isDefined) {
+      val start = System.nanoTime()
+      while (start + sleep.get >= System.nanoTime()) { }
+    }
+
+    // fail at the end if requested
+    if (fail && !it.hasNext) throw new RuntimeException()
+
+    // just iterate
+    if (fail) {
+      -it.next()
+    } else {
+      it.next()
+    }
+  }
+}
+
+// we fail task two 3 times and delay task four so we see a speculative execution
+val result = (
+  spark.range(0, 10000000, 1, 100)
+       .mapPartitions { it => {
+         val ctx = TaskContext.get()
+         FaultyIterator(
+           it,
+           ctx.partitionId == 2 && ctx.attemptNumber < 3,
+           Some(ctx.partitionId == 4).filter(v => v).map(_ => 250000)
+         )
+       }}
+       .groupBy(($"value" / 1000000).cast("int"))
+       .as[Long, Long]
+       .mapGroups{(id, it) => (id, it.length)}
+       .sort("_1")
+       .collect
+  )
+```
+
+We can compare the result with the expected outcome:
+```Scala
+assert(result.sameElements(Array((0,1000000), (1,1000000), (2,1000000), (3,1000000), (4,1000000), (5,1000000), (6,1000000), (7,1000000), (8,1000000), (9,1000000))))
+```
+
+## Stop the docker cluster
+
+Finally, stop the cluster:
+
+```bash
+docker compose -f deploy/docker-compose/docker-compose.yml down
+```
+```
+[+] Running 12/12
+ ✔ Container rss-shuffle-server-1  Removed                                                    10.5s
+ ✔ Container rss-shuffle-server-2  Removed                                                    10.7s
+ ✔ Container rss-shuffle-server-3  Removed                                                    10.5s
+ ✔ Container rss-shuffle-server-4  Removed                                                    10.6s
+ ✔ Container rss-spark-worker-1    Removed                                                     0.8s
+ ✔ Container rss-spark-worker-2    Removed                                                     1.0s
+ ✔ Container rss-spark-worker-3    Removed                                                     0.9s
+ ✔ Container rss-spark-worker-4    Removed                                                     1.1s
+ ✔ Container rss-spark-master-1    Removed                                                     1.6s
+ ✔ Container rss-coordinator-1     Removed                                                    10.4s
+ ✔ Container rss-coordinator-2     Removed                                                    10.5s
+ ✔ Network rss_default             Removed                                                     0.4s
+```
+
+docker exec -it rss-spark-master-1 /opt/spark/bin/spark-shell \
```

Comment:
Forgot adding a new title? This paragraph looks strange.

Reply:
fixed